gadget/update.go

```diff
@@ -326,6 +327,95 @@ func EnsureLayoutCompatibility(gadgetLayout *LaidOutVolume, diskLayout *OnDiskVo
 	return nil
 }
 
+// DiskVolumeDeviceTraitsForDevice takes a gadget volume and an expected
+// disk device path and builds up the traits for that device. It validates that
+// the specified volume can be laid out exactly to the disk device, returning an
+// error if the two are not compatible.
+func DiskVolumeDeviceTraitsForDevice(laidOutVolume *LaidOutVolume, dev string) (res DiskVolumeDeviceTraits, err error) {
+	vol := laidOutVolume.Volume
+
+	// get the disk layout for this device
+	diskLayout, err := OnDiskVolumeFromDevice(dev)
+	if err != nil {
+		return res, fmt.Errorf("cannot read %v partitions for candidate volume %s: %v", dev, vol.Name, err)
+	}
+
+	// ensure that the on disk and the laid out version are actually compatible
+	opts := &EnsureLayoutCompatibilityOptions{
+		AssumeCreatablePartitionsCreated: true,
+	}
+	if err := EnsureLayoutCompatibility(laidOutVolume, diskLayout, opts); err != nil {
```

Comment:
This check feels a bit out of place here. It's not clear why a function that just returns the disk partitions should check the compatibility with another layout which this function has no business with (it's not applying it, as far as I can see).

I would actually remove it altogether, and let the caller do the check (so we also solve the naming issue :-) ).